test/pool.ts

```diff
@@ -0,0 +1,158 @@
+/**
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+import {use, expect} from 'chai';
+import * as chaiAsPromised from 'chai-as-promised';
+
+import {ClientPool} from '../src/pool';
+import {Deferred} from './util/helpers';
+
+use(chaiAsPromised.default);
+
+function deferredPromises(count: number): Array<Deferred<void>> {
+  const deferred: Array<Deferred<void>> = [];
+  for (let i = 0; i < count; ++i) {
+    deferred.push(new Deferred<void>());
+  }
+  return deferred;
+}
+
+describe('Client pool', () => {
+  it('creates new instances as needed', () => {
+    const clientPool = new ClientPool<{}>(3, () => {
+      return {};
+    });
+
+    expect(clientPool.size).to.eq(0);
+
+    const deferred = deferredPromises(4);
+
+    clientPool.run(() => deferred[0].promise);
```

Comment:
Very optional: perhaps use a `for` loop here (and in similar places below)?

Reply:
~~Uhm. Yes, of course.~~

I just realized that this makes the tests harder to follow since each loop iteration would have to assert a slightly different state. For that reason, I left it as is.